wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -1085,10 +1087,91 @@ private class WasmExpressionBuilder private (
             genIsPrimType(primType)
           case None =>
             val info = ctx.getClassInfo(testClassName)
+
+            // TODO: Efficient type inclusion test
+            // Current implementation walk through the itables of the Object which takes O(N)
+            // (where N = number of interfaces the expr implements)
+            // Also, this procedure can't be extracted as a helper function in Wasm because the
+            // immediate argument of `ref.test` is specific to the `testClassName`,
+            // which result in a number of instructions everywhere at `isInstanceOf[interface]`.
+            // See: https://github.com/tanishiking/scala-wasm/issues/27#issuecomment-2008252049
             if (info.isInterface) {
-              // TODO: run-time type test for interface
-              println(tree)
-              ???
+              val expr = fctx.addLocal(
+                fctx.genSyntheticLocalName(),
+                TypeTransformer.transformType(IRTypes.AnyType)(ctx)
+              )
+              val found = fctx.addLocal(fctx.genSyntheticLocalName(), Types.WasmInt32)
+              val cnt = fctx.addLocal(fctx.genSyntheticLocalName(), Types.WasmInt32)
+              val len = fctx.addLocal(fctx.genSyntheticLocalName(), Types.WasmInt32)
+              val itables = fctx.addLocal(
+                fctx.genSyntheticLocalName(),
+                Types.WasmRefNullType(Types.WasmHeapType.Type(WasmArrayType.itables.name))
+              )
+
+              instrs += LOCAL_SET(expr)
+              instrs += I32_CONST(I32(0))
+              instrs += LOCAL_SET(found)
+
+              fctx.block() { testFail =>
+                // if expr is not an instance of Object, return false
+                instrs += LOCAL_GET(expr)
+                instrs += REF_TEST(HeapType(Types.WasmHeapType.ObjectType))
+                instrs += I32_CONST(I32(1))
+                instrs += I32_XOR
+                instrs += BR_IF(testFail)
```

Comment:
Is this the same as `BR_IF_CAST_FAIL` here?

Reply:
fixed https://github.com/tanishiking/scala-wasm/pull/29/commits/ba4858b6f672acf0463c11a02f11adc81a39df28